### PR TITLE
Triplet loss

### PR DESCRIPTION
**Pull request by Ghostvv**
**Proposed changes**:
- reformulate our margin loss as triplet loss

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [x] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by Ghostvv** [comment]
/modeltest

```yml
 dataset_branch: "triplet-loss"
 include:
  - dataset: ["all"]
    config: ["all"]
```

File: rasa/utils/tensorflow/layers.py [hunk 1]
```diff
@@ -806,38 +801,32 @@ def _loss_margin(
...
         if self.use_max_sim_neg:
-            # minimize only maximum similarity over incorrect actions
+            # minimize only maximum similarity over incorrect labels
             max_sim_neg_il = tf.reduce_max(sim_neg_il, axis=-1)
             loss += tf.maximum(0.0, self.mu_neg + max_sim_neg_il)
         else:
-            # minimize all similarities with incorrect actions
+            # minimize all similarities with incorrect labels
             max_margin = tf.maximum(0.0, self.mu_neg + sim_neg_il)
             loss += tf.reduce_sum(max_margin, axis=-1)
```
Comment (dakshvar22) [inline rasa/utils/tensorflow/layers.py:820]: This should be `reduce_mean` to avoid extra negative terms as we discussed on slack.

Comment (Ghostvv) [reply rasa/utils/tensorflow/layers.py:820]: good point!

Comment (Ghostvv) [inline rasa/utils/tensorflow/layers.py:820]:
```suggestion
            loss += tf.reduce_mean(max_margin, axis=-1)
```

File: docs/docs/components.mdx [hunk 2]
```diff
@@ -1542,7 +1542,7 @@ However, additional parameters exist that can be adapted.
 |                                 |                  | or 'inner'.                                                  |
 +---------------------------------+------------------+--------------------------------------------------------------+
 | loss_type                       | "cross_entropy"  | The type of the loss function, either 'cross_entropy'        |
-|                                 |                  | or 'margin'.                                                 |
+|                                 |                  | or 'triplet'.                                                |
```
Comment (dakshvar22) [inline docs/docs/components.mdx:1545]: Since margin loss is still in the codebase, we should have it in the documentation too (can mention deprecated for that option). This topic came up when we were deprecating components like `HFTransformersNLP` and the recommendation was to only remove it once it is removed from the codebase.
Same for other parameters below.

File: rasa/nlu/classifiers/diet_classifier.py [hunk 3]
```diff
@@ -177,7 +177,7 @@ def required_components(cls) -> List[Type[Component]]:
         NUM_NEG: 20,
         # Type of similarity measure to use, either 'auto' or 'cosine' or 'inner'.
         SIMILARITY_TYPE: AUTO,
-        # The type of the loss function, either 'cross_entropy' or 'margin'.
+        # The type of the loss function, either 'cross_entropy' or 'triplet'.
```
Comment (dakshvar22) [inline rasa/nlu/classifiers/diet_classifier.py:180]: Can you make this change in response selector also please?

File: rasa/utils/train_utils.py [hunk 4]
```diff
@@ -557,12 +560,32 @@ def _check_loss_setting(component_config: Dict[Text, Any]) -> None:
...
-            category=UserWarning,
+            warn_until_version=NEXT_MAJOR_VERSION_FOR_DEPRECATIONS,
+        )
+    if component_config.get(NEGATIVE_MARGIN_SCALE) is not None:
+        rasa.shared.utils.io.raise_deprecation_warning(
+            f"{NEGATIVE_MARGIN_SCALE} is deprecated. It will be removed in "
+            f"Rasa Open Source 3.0.0 onwards.",
+            warn_until_version=NEXT_MAJOR_VERSION_FOR_DEPRECATIONS,
+        )
+    if component_config.get(NEGATIVE_MARGIN_SCALE) is not None:
```
Comment (dakshvar22) [inline rasa/utils/train_utils.py:576]:
```suggestion
    if component_config.get(USE_MAX_NEG_SIM) is not None:
```

File: rasa/utils/train_utils.py [hunk 5]
```diff
@@ -557,12 +560,32 @@ def _check_loss_setting(component_config: Dict[Text, Any]) -> None:
         SOFTMAX,
         CROSS_ENTROPY,
     ]:
-        rasa.shared.utils.io.raise_warning(
+        rasa.shared.utils.io.raise_deprecation_warning(
             f"{CONSTRAIN_SIMILARITIES} is set to `False`. It is recommended "
             f"to set it to `True` when using cross-entropy loss. It will be set to "
```
Comment (dakshvar22) [inline rasa/utils/train_utils.py:565]: You'll need to update these functions too - 
`update_similarity_type`
`_check_similarity_loss_setting`

Also set `similarity_type` to `cosine` if `loss_type=triplet` and `similarity_type=auto`.

**Review by dakshvar22 — CHANGES_REQUESTED** [review]
Can you please also modify tests to add test cases with `loss_type=triplet` in addition to `loss_type=margin`?

File: rasa/utils/train_utils.py [hunk 6]
```diff
@@ -483,14 +486,16 @@ def update_confidence_type(component_config: Dict[Text, Any]) -> Dict[Text, Any]
...
             f"Overriding defaults by setting {MODEL_CONFIDENCE} to "
-            f"{AUTO} as {LOSS_TYPE} is set to {MARGIN} in the configuration. This means that "
-            f"model's confidences will be computed as cosine similarities. "
-            f"Users are encouraged to shift to cross entropy loss by setting `{LOSS_TYPE}={CROSS_ENTROPY}`."
+            f"{AUTO} as {LOSS_TYPE} is set to {component_config[LOSS_TYPE]} "
+            f"in the configuration. "
+            f"This means that model's confidences will be computed as "
+            f"cosine similarities. "
         )
         component_config[MODEL_CONFIDENCE] = AUTO
```
Comment (Ghostvv) [inline rasa/utils/train_utils.py:497]: @dakshvar22 I didn't find where to change `AUTO`. Do I keep it like that?

**Review by dakshvar22 — COMMENTED** [review]
Just one comment about the warning. Can that be fixed somehow?
